kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bytes-simplification.k

```diff
@@ -75,17 +75,27 @@ module BYTES-SIMPLIFICATION [symbolic]
     rule [buf-asWord-invert-rl]:
       #asWord ( #buf ( WB:Int, X:Int ) ) => X
       requires 0 <=Int WB andBool 0 <=Int X andBool
-               ( ( WB <=Int 32 andBool X <Int 2 ^Int (WB *Int 8) ) orBool
+               ( ( WB <=Int 32 andBool X <Int 2 ^Int (8 *Int WB) ) orBool
                  ( WB  >Int 32 andBool X <Int pow256 ) )
       [simplification, concrete(WB), preserves-definedness]
 
     rule [buf-asWord-invert-rl-range]:
       #asWord ( #range ( #buf ( WB:Int, X:Int ), S:Int, WR:Int ) ) => X
       requires 0 <=Int WB andBool 0 <=Int X andBool 0 <=Int S andBool 0 <=Int WR
        andBool S +Int WR ==Int WB
-       andBool X <Int 2 ^Int (WR *Int 8)
+       andBool X <Int 2 ^Int (8 *Int WR)
       [simplification, concrete(WB, S, WR), preserves-definedness]
 
+    rule [asWord-trim-leading-zeros-concat]:
+      #asWord ( BZ +Bytes B ) => #asWord ( B )
+      requires #asInteger ( BZ ) ==Int 0
+      [simplification, concrete(BZ)]
+
+    rule [asWord-trim-overflowing-zeros]:
+      #asWord ( B ) => #asWord ( #range(B, lengthBytes(B) -Int 32, 32) )
+      requires 32 <Int lengthBytes(B) andBool #asInteger(#range(B, 0, lengthBytes(B) -Int 32)) ==Int 0
+      [simplification]
```

Comment:
Is the second condition here actually necessary? I think it doesn't matter if the higher bits are set, `#asWord(...)` will trim them away anyway, so we just need `32 <Int lengthBytes(B)` to prevent infinite loop, right?

Reply:
For this particular lemma, yes, otherwise `lengthBytes(B) -Int 32 <Int 0`. I just wanted to have this lemma as well to contain the buffer size to max 32 whenever possible. There are others that will do the cropping in the `#buf ( X, #asWord ( ...` scenario, when we need to fit the number into a specified `X` bytes.

Reply:
Sorry, misunderstood initially - no, it's not needed at all.

Reply:
Removed now.